File: 1029-vertical-order-traversal-of-a-binary-tree/vertical-order-traversal-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
       
        map<int, map<int, multiset<int>>> nodes;
        queue<pair<TreeNode*, pair<int, int>>> q;

        if (root == NULL) return {};

        q.push({root, {0, 0}}); // {node, {hd, level}}

        while (!q.empty()) {
            auto temp = q.front(); q.pop();
            TreeNode* front = temp.first;
            int hd = temp.second.first;
            int level = temp.second.second;

            nodes[hd][level].insert(front->val); // use multiset to keep values sorted

            if (front->left) 
                q.push({front->left, {hd - 1, level + 1}});
            if (front->right) 
                q.push({front->right, {hd + 1, level + 1}});
        }

        vector<vector<int>> ans;
        for (auto& p : nodes) {
            vector<int> col;
            for (auto& q : p.second) {
                col.insert(col.end(), q.second.begin(), q.second.end());
            }
            ans.push_back(col);
        }
        return ans;
    }
};

File: 1029-vertical-order-traversal-of-a-binary-tree/vertical-order-traversal-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Pair{
    TreeNode node;
    int v;
    int level;
    Pair(TreeNode node,int v,int level){
        this.node=node;
        this.v=v;
        this.level=level;
    }
}
class Solution {
    public List<List<Integer>> verticalTraversal(TreeNode root) {
        TreeMap<Integer,TreeMap<Integer,PriorityQueue<Integer>>> nodes= new TreeMap<>();
        Queue<Pair> q= new LinkedList<>();
        
        if(root==null) return new ArrayList<>();
        q.add(new Pair(root,0,0));

        while(!q.isEmpty()){
            Pair p= q.poll();
            TreeNode node= p.node;
            int col= p.v;
            int row= p.level;

            nodes.putIfAbsent(col, new TreeMap<>());
            nodes.get(col).putIfAbsent(row,new PriorityQueue<>());

            nodes.get(col).get(row).add(node.val);

            if(node.left!= null)
                q.add(new Pair(node.left, col-1, row+1));
            if(node.right!= null)
                q.add(new Pair(node.right, col+1, row+1));
        }

        List<List<Integer>> ans= new ArrayList<>();
        for(TreeMap<Integer, PriorityQueue<Integer>> rows : nodes.values()){
            List<Integer> column= new ArrayList<>();
            for(PriorityQueue<Integer> pq: rows.values()){
                while(!pq.isEmpty()){
                    column.add(pq.poll());
                }
            }
            ans.add(column);
        }
        return ans;
    }
}